falcon/app_helpers.py

```diff
@@ -52,9 +56,9 @@ def prepare_middleware(middleware, independent_middleware=False, asgi=False):
 
     # PERF(kgriffs): do getattr calls once, in advance, so we don't
     # have to do them every time in the request path.
-    request_mw = []
-    resource_mw = []
-    response_mw = []
+    request_mw: List = []
+    resource_mw: List = []
+    response_mw: List = []
```

Comment:
Here too we should use List[Any]. Doing just list is not really telling much to the type checker and it may complain that a list is missing the generic parameter

Reply:
Not sure I understand the reasoning to do so tbh... :thinking: 

According to https://docs.python.org/3/library/typing.html#user-defined-generic-types:
> Using a generic class without specifying type parameters assumes [Any](https://docs.python.org/3/library/typing.html#typing.Any) for each position. In the following example, `MyIterable` is not generic but implicitly inherits from `Iterable[Any]`:
```python
from collections.abc import Iterable

class MyIterable(Iterable): # Same as Iterable[Any]
    ...
```

This is for subclassing, but I would assume the checker interprets signatures in the same way?

Reply:
Depending on configuration, mypy will complain.

The following snippet
```py
from typing import Iterable

def foo(x: Iterable) -> None:
    print(list(x))
```
reports
```
error: Missing type parameters for generic type "Iterable"  [type-arg]
```

Like I commented we can ignore this for now, and revisit once we have more typing in falcon

Reply:
Aha, I see, this is what [`--disallow-any-generics`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-disallow-any-generics) does... Would this explode for the user even if this was inside Falcon, or are packages used not checked the same way?

Reply:
Not sure actually if it effects libraries.

Yes the fact that you have to use the upper case collections is annoying.

Reply:
also note that I think we can keep the implicit generics, since what we have here is better than nothing.

We can improve later on!

Reply:
Indeed, I think it is not a problem for now.

I have tested this, and `mypy` seems to accept it:
```python
import falcon


def hello_world(req, resp):
    resp.media = {'msg': 'Hello, World!'}


def create_app() -> falcon.App:
    app = falcon.App()
    app.add_sink(hello_world)

    return app


if __name__ == '__main__':
    app = create_app()
    print(app)
```

```
mypy --disallow-any-generics test.py 
Success: no issues found in 1 source file
```

`add_sink()` has a generic `Callable`, and if you try typing Falcon itself with `--disallow-any-generics`, it complains. But it doesn't complain when just using Falcon. If I tried passing an `int` for sink pattern, for instance, it does correctly detect the error.

Reply:
That's great, thanks for testing